src/coreclr/src/jit/morph.cpp

```diff
@@ -18213,6 +18319,57 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
         }
     }
 
+    //------------------------------------------------------------------------
+    // MorphLocalAddress: Change a tree that represents a local variable address
+    //    to a single LCL_VAR_ADDR or LCL_FLD_ADDR node.
+    //
+    // Arguments:
+    //    val - a value that represents the local address
+    //
+    void MorphLocalAddress(const Value& val)
+    {
+        assert(val.IsAddress());
+        assert(val.Node()->TypeIs(TYP_BYREF, TYP_I_IMPL));
+        assert(m_compiler->lvaVarAddrExposed(val.LclNum()));
```

Comment:
I know that the JIT conflates the notions of a local being address-taken and address-exposed, but I think that's something we need to fix - migrating `lvHasLdAddrOp` to `lvAddrTaken` and leaving `lvAddrExposed` to actually mean that the address has potentially been exposed beyond this method. In that vein I think that this assert could allow either to be true.

Reply:
> and leaving lvAddrExposed to actually mean that the address has potentially been exposed beyond this method

Hmm? While that may be useful in some cases, the JIT doesn't currently have such capability and it's not clear what it would take to add it. `LocalAddressVisitor` can detect when a local address is passed as a call argument within the same tree but it currently doesn't not track values across trees. It's likely that it can be extended to do that as well - on assignment, the  `Value` computed on the RHS could be associated with the destination local variable so we can continue tracking address uses, at least within the block if not the entire method. But such work seems rather orthogonal to `ADDR` elimination.

> I know that the JIT conflates the notions of a local being address-taken and address-exposed

The goal is to make address-taken and address-exposed equivalent:
* Any use of `LCL_VAR|FLD_ADDR` on a variable => variable is address taken/exposed and thus not tracked
* Any tree that represents an indirect access to a variable (well, any tree that `LocalAddressVisitor` can recognize as such) is converted to a single `LCL_VAR|FLD` node that represents an equivalent access. The recognized trees are expected to be a superset of the tress recognized by `IsLocalAddrExpr` & co. so this would result `IsLocalAddrExpr` no longer being necessary.
* To put it another way - liveness & SSA will only work with `LCL_VAR|FLD` and never with `LCL_VAR|FLD_ADDR`.

Reply:
So ... it may be that ultimately there's no disadvantage to making these equivalent. That is, if we postpone any marking of `lvAddrExposed` until `Lowering` when an address is required for code generation, then we presumably won't unduly pessimize the handling of address-taken (as opposed to address-exposed) locals.

Reply:
> That is, if we postpone any marking of lvAddrExposed until Lowering when an address is required for code generation

We can't, `lvAddrExposed` has to be set before liveness in order to determine if a variable is tracked or not. In fact, it likely needs to be set even before global morph due to local assertion propagation.

>  then we presumably won't unduly pessimize the handling of address-taken (as opposed to address-exposed) locals

Ah, but there's no pessimization going on here, except perhaps the corner case of variables having field accesses with an offset larger than 65535. The "address taken but not address exposed" case does not transform into "address exposed", the case simply disappears.

Reply:
Perhaps I wasn't clear. I believe that we want to track/optimize structs that are only address-taken due to code generation requirements where it needs the value to be on the stack, but are not actually address-exposed.
Or maybe you're thinking that for those we never mark them `lvAddrExposed`? If so, how do we indicate to the register allocator and the frame layout code that they must be on the stack?

Reply:
> Or maybe you're thinking that for those we never mark them lvAddrExposed?

Yes, these should not be `lvAddrExposed`.

> If so, how do we indicate to the register allocator and the frame layout code that they must be on the stack?

Isn't that what `lvDoNotEnregister` is for?

BTW, I think it is the setting of `lvDoNotEnregister` that should be postponed until lowering. Currently we're setting it as soon as we create `LCL_FLD`s in morph but I think that eventually `LCL_FLD`s should not prevent enregistration by default. But I haven't yet checked if there's anything else before register allocation that would really need `lvDoNotEnregister`.

Reply:
Okay, that makes sense. I agree about postponing the setting of `lvDoNotEnregister` as much as possible. Although I haven't looked in detail, I believe that most of the cases where we set it are due to struct copies, argument passing and returns - and IMO we shouldn't even create `LCL_FLD`s for any of those until `Lowering` (that is, unless they're actually fields, as opposed to actually being "reinterpret casts") and it should also be responsible for determine when these contexts require the value to be on the stack.

Reply:
> and IMO we shouldn't even create LCL_FLDs for any of those until Lowering (that is, unless they're actually fields, as opposed to actually being "reinterpret casts") and it should also be responsible for determine when these contexts require the value to be on the stack.

Yes, I hope that this ADDR/ASG and 1st class struct saga will also lead me to deal with some reinterpret casts. And I'd also love to see most ABI handling moved to lowering, though it's not very clear to me what exactly can be moved.

There's seems to be a rather convoluted trade-off between ABI details that are not relevant to optimizations and those that are. For example - if a single float struct has to be passed in a GPR we need a `movd`. If the same struct values gets passed to 2 call arguments, does that mean that we'd want the `movd` to be created early (as `BITCAST`) so it can be picked up by CSE? Is it even worth considering such cases? While I'm convinced that the current JIT can be improved I don't think it should handle every possible case.

Speaking of variable enregistration, I have a question about LSRA: Is there anything that would force an enregistered variable to be spilled to memory? Kind of the opposite of `RegOptional`.

Reply:
I'm not sure that the value of CSE'ing `BITCAST`s would justify supporting it. While it wouldn't be complex, it would require, as you point out, that we expose those ABI details early. That said, I think we also need to continue to consider addressing modes early on (though instead of rediscovering them in `Lowering` I'd like to see us actually create them early), which requires understanding the available addressing modes on the target. So there's clearly value in having some abstraction of some target details exposed early in the JIT. But for the ABI stuff, I think we're best off moving as much as possible to `Lowering`, and then perhaps deciding later what to reintroduce early.

> Is there anything that would force an enregistered variable to be spilled to memory? Kind of the opposite of RegOptional

If you mean something that would say "spill this to memory here", then no, there's nothing quite like that currently. The "writeThru" that I'm working on for EH-exposed variables is kind of like that, though it is applied to all defs. I also hope to build on it to enable the spilling of single-def variables while still keeping it live in a register while possible.